datafusion/catalog/src/table.rs

```diff
@@ -297,3 +299,40 @@ pub trait TableProviderFactory: Debug + Sync + Send {
         cmd: &CreateExternalTable,
     ) -> Result<Arc<dyn TableProvider>>;
 }
+
+/// A trait for table function implementations
+pub trait TableFunctionImpl: Debug + Sync + Send {
```

Comment:
```suggestion
/// Returns this function's name
    fn name(&self) -> &str;
```

This is much more consistent with other UDFs

Reply:
This is the part of the refactor, changing this requires several fixes. We can leave it to a separate PR to make this PR smaller

Reply:
I also recommend to take `name()` inside `TableFunctionImpl`

Reply:
I think doing it as a follow on PR makes sense to me -- perhaps we can file a ticket

Another thing that would be nice for a follow on PR is to move this trait into its own module (`catalog/src/table_function.rs` perhaps)

Reply:
Filed https://github.com/apache/datafusion/issues/13613